File: Algorithms/008_StringToInteger(atoi).cpp
#include "leetcode.h"
class Solution {
public:
    int atoi(string str) {
        int len = str.size();
        int i = 0;
        bool positive = true;
        long long result = 0;
        while (i < len && isspace(str[i])) 
            i ++;
        if (i < len && str[i] == '+') 
            i ++;
        else if (i < len && str[i] == '-') {
            positive = false;
            i ++;
        }
        while (i < len) {
            if (isdigit(str[i])) {
                if (result > 214748364 || (result == 214748364 && (str[i] - '0') >= 8)) {
                    return positive? 2147483647:-2147483648;
                }
                result = 10*result + (str[i] - '0');
            } else {
                break;
            }
            i ++;
        }
        if (positive)
            return result;
        else 
            return -result;
    }
};
int main() {
    Solution s;
    cout << s.atoi(" 1.3") << endl;
    cout << s.atoi("9223372036854775809") << endl;
    return 0;
}

File: Algorithms/008.StringToInteger(atoi).py
class Solution:
    # @return an integer
    def atoi(self, str):
        length = len(str)
        i = 0
        positive = True
        result = 0
        while i < length and str[i].isspace():
            i += 1
        if i < length and str[i] == '+':
            i += 1
        elif i < length and str[i] == '-':
            positive = False
            i += 1
        while i < length:
            if str[i].isdigit():
                if result > 214748364 or (result == 214748364 and (ord(str[i]) - ord('0')) >= 8):
                    return 2147483647 if positive else -2147483648
                result = 10*result + (ord(str[i]) - ord('0'))
            else:
                break
            i += 1
        if positive:
            return result
        else:
            return -result
